### DISCUSSION TITLE
Why I can't bind StyledProperty in TemplatedControl?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16772] [Q&A] [answered]
Here is my minimum project:
```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
					xmlns:Local="using:AvaloniaApplication1"
        xmlns:controls="using:AvaloniaApplication1">
 
  <ControlTheme x:Key="{x:Type controls:TemplatedControl1}" TargetType="{x:Type controls:TemplatedControl1}">	  
	  <Style Selector="^ /template/ TextBlock.Test">
		  <Setter Property="Foreground" Value="{TemplateBinding HoverForeground}"></Setter>
	  </Style>
	  <Setter Property="Template">
		  <Setter Property="HoverForeground" Value="Red"></Setter>
		  <ControlTemplate>
			  <Grid>
				  <TextBlock Classes="Test">123123</TextBlock>
			  </Grid>
		  </ControlTemplate>
	  </Setter>
  </ControlTheme>
</ResourceDictionary>
```

And here is code-behind:
```
using Avalonia.Controls;
using Avalonia;
using Avalonia.Controls.Primitives;
using Avalonia.Media;

namespace AvaloniaApplication1;

public class TemplatedControl1 : TemplatedControl
{
    
    public static readonly StyledProperty<IBrush> HoverForegroundProperty =
AvaloniaProperty.Register<TemplatedControl1, IBrush>(nameof(HoverForeground));

    public IBrush HoverForeground
    {
        get { return GetValue(HoverForegroundProperty); }
        set { SetValue(HoverForegroundProperty, value); }
    }  
}
```

When the project compiled, VS reports this error:

> Severity	Code	Description	Project	File	Line	Suppression State	Details
> Error		Unable to find a setter that allows multiple assignments to the property Value of type Avalonia.Base:Avalonia.Styling.Setter Line 8, position 5.	AvaloniaApplication1	C:\Users\Admin\Desktop\AvaloniaApplication1\TemplatedControl1.axaml	8		

What's going on?

What I actually did was use Style to control the Foreground of several controls at the same time to make it consistent with HoverForeground.

**Comment by rabbitism** [comment]
Why do you have a setter in setter?

**Reply by kyuranger** [reply]
The setter
`<Setter Property="HoverForeground" Value="Red"></Setter>`
is to set the default value of the HoverForeground.

The setter
`<Setter Property="Foreground" Value="{TemplateBinding HoverForeground}"></Setter>`
is to bind the value of the foreground to HoverForeground.

I just tried deleted `<Setter Property="HoverForeground" Value="Red"></Setter>`
and it still gives an error.

**Reply by kyuranger** [reply]
BTW, I was familiar with using TemplatedControl in Style form in Avalonia before, but I cannot use it anymore since it was changed to TemplatedControl in ControlTheme form in a certain version. There are not many tutorials about TemplatedControl in ControlTheme form.

**Reply by rabbitism** [reply]
That must be another error. What is the new error?

**Reply by kyuranger** [reply]
There are two errors:

> Severity	Code	Description	Project	File	Line	Suppression State	Details
> Error (active)	AVLN:0004	Unable to resolve suitable regular or attached property HoverForeground on type Avalonia.Controls:Avalonia.Controls.TextBlock Line 8, position 35.	AvaloniaApplication1	C:\Users\Admin\Desktop\AvaloniaApplication1\TemplatedControl1.axaml	8		
> 

And

> Severity	Code	Description	Project	File	Line	Suppression State	Details
> Error		Cannot parse a compiled binding without an explicit x:DataType directive to give a starting data type for bindings. Line 8, position 35.	AvaloniaApplication1	C:\Users\Admin\Desktop\AvaloniaApplication1\TemplatedControl1.axaml	8

**Comment by rabbitism** [comment, marked as answer]
First of all you can always define ControlTemplate in Style instead of ControlTheme, that is always supported. 
For your errors. TemplateBinding should not be used like this. You can use full template parent binding syntax like below. 
```xml
<Setter Property="Foreground" Value="{Binding Path=HoverForeground, RelativeSource={RelativeSource TemplatedParent}}"></Setter>
```

**Reply by kyuranger** [reply]
Thank you. It works!
